File: examples/LMemLoopback/client/go/AdvancedStatic/LMemLoopbackClient.go
package main

import (
    "../gen-go/com/maxeler/LMemLoopback"
    "fmt"
    "git.apache.org/thrift.git/lib/go/thrift"
    "os"
    "time"
)

func checkForErrors(err error){
    if err != nil {
        fmt.Fprintln(os.Stderr, err)
        os.Exit(-1)
    }
}

func check(dataOutDFE []int32, dataOutCPU []int32, size int64) (int64) {
    status := int64(0);

    for i := int64(0); i < size; i++ {
        if dataOutDFE[i] != dataOutCPU[i] {
            fmt.Printf("Output data @ %v = %v (expected %v)\r\n", i, dataOutDFE[i], dataOutCPU[i])
            status++
        }
    }

    return status;
}

// Calculates time passed from startTime in seconds
func calcTime(startTime time.Time) (float64){
    return float64(time.Now().UnixNano() - startTime.UnixNano()) / 1000000000
}

// LMemLoopbackCPU implementation
func LMemLoopbackCPU(size int64, inA []int32, inB []int32) ([]int32) {
    dataOut := make([]int32, size)

    for i := int64(0) ; i < size ; i++ {
        dataOut[i] = inA[i] + inB[i]
    }

    return dataOut
}

// LMemLoopbackDFE AdvancedStatic implementation
func LMemLoopbackDFE(size int64, inA []int32, inB []int32) ([]int32){
    outData := make([]int32, size)
    sizeBytes := size * 4

    startTime := time.Now()
  
    // Make socket
    transport, err := thrift.NewTSocket("localhost:9090"); checkForErrors(err)

    // Wrap in a protocol	
    protocol := thrift.NewTBinaryProtocolFactoryDefault()

    // Create a client to use the protocol encoder
    client := LMemLoopback.NewLMemLoopbackServiceClientFactory(transport, protocol)

    currentTime := calcTime(startTime)
    fmt.Printf("Creating a client:\t\t\t\t%.5fs\n", currentTime)

    // Connect!
    startTime = time.Now()
    err = transport.Open(); checkForErrors(err)
    currentTime = calcTime(startTime)
    fmt.Printf("Opening connection:\t\t\t\t%.5fs\n", currentTime)

    // Allocate and send input streams to server
    startTime = time.Now()
    addressInA, err := client.MallocInt32T(size); checkForErrors(err)
    client.SendDataInt32T(addressInA, inA)

    addressInB, err := client.MallocInt32T(size); checkForErrors(err)
    client.SendDataInt32T(addressInB, inB);
    currentTime = calcTime(startTime)
    fmt.Printf("Sending input data:\t\t\t\t%.5fs\n", currentTime)

    // Initialize maxfile
    startTime = time.Now()
    maxfile, err := client.LMemLoopbackInit(); checkForErrors(err)
    currentTime = calcTime(startTime)
    fmt.Printf("Initializing maxfile:\t\t\t\t%.5fs\n", currentTime)

    // Load DFE
    startTime = time.Now()
    engine, err := client.MaxLoad(maxfile, "*"); checkForErrors(err)
    currentTime = calcTime(startTime)
    fmt.Printf("Loading DFE:\t\t\t\t\t%.5fs\n", currentTime)

    // Allocate memory for output stream on server
    startTime = time.Now()
    addressOutData, err := client.MallocInt32T(size); checkForErrors(err)
    currentTime = calcTime(startTime)
    fmt.Printf("Allocating memory for output stream on server:\t%.5fs\n", currentTime)

    // Write to LMem
    startTime = time.Now()
    actionsLmemWrite := LMemLoopback.NewLMemLoopbackWriteLMemActionsTStruct()

    actionsLmemWrite.ParamAddress = 0
    actionsLmemWrite.ParamNbytes = sizeBytes
    actionsLmemWrite.InstreamCpuToLmem = addressInA

    addressActionsLmemWrite, err := client.Send_LMemLoopbackWriteLMemActionsT(actionsLmemWrite); checkForErrors(err)
    err = client.LMemLoopbackWriteLMemRun(engine, addressActionsLmemWrite); checkForErrors(err)

    actionsLmemWrite.ParamAddress = sizeBytes
    actionsLmemWrite.ParamNbytes = sizeBytes
    actionsLmemWrite.InstreamCpuToLmem = addressInB

    addressActionsLmemWrite, err = client.Send_LMemLoopbackWriteLMemActionsT(actionsLmemWrite); checkForErrors(err)
    err = client.LMemLoopbackWriteLMemRun(engine, addressActionsLmemWrite); checkForErrors(err)
    client.Free(addressActionsLmemWrite);
    currentTime = calcTime(startTime)
    fmt.Printf("Writing to LMem:\t\t\t\t%.5fs\n", currentTime)

    // Action default
    startTime = time.Now()
    actions := LMemLoopback.NewLMemLoopbackActionsTStruct();

    actions.Param_N = int32(size)

    addressActions, err := client.Send_LMemLoopbackActionsT(actions); checkForErrors(err)
    err = client.LMemLoopbackRun(engine, addressActions); checkForErrors(err)
    client.Free(addressActions)
    currentTime = calcTime(startTime)
    fmt.Printf("LMemLoopback time:\t\t\t\t%.5fs\n", currentTime)

    // Read from LMem
    startTime = time.Now()
    actionsLmemRead := LMemLoopback.NewLMemLoopbackReadLMemActionsTStruct()

    actionsLmemRead.ParamAddress = 2 * sizeBytes
    actionsLmemRead.ParamNbytes = sizeBytes
    actionsLmemRead.OutstreamLmemToCpu = addressOutData

    addressActionsLmemRead, err := client.Send_LMemLoopbackReadLMemActionsT(actionsLmemRead); checkForErrors(err)
    err = client.LMemLoopbackReadLMemRun(engine, addressActionsLmemRead); checkForErrors(err)
    client.Free(addressActionsLmemRead)
    currentTime = calcTime(startTime)
    fmt.Printf("Reading from LMem:\t\t\t\t%.5fs\n", currentTime)

    // Unload DFE
    startTime = time.Now()
    err = client.MaxUnload(engine); checkForErrors(err)
    currentTime = calcTime(startTime)
    fmt.Printf("Unloading DFE:\t\t\t\t\t%.5fs\n", currentTime)

    // Get output stream from server
    startTime = time.Now()
    outData, err = client.ReceiveDataInt32T(addressOutData, size); checkForErrors(err)
    currentTime = calcTime(startTime)
    fmt.Printf("Getting output stream:\t(size = %v bit)\t%.5fs\n", size * 32, currentTime)

    // Free allocated memory for streams on server
    startTime = time.Now()
    client.Free(addressInA)
    client.Free(addressInB)
    client.Free(addressOutData)
    currentTime = calcTime(startTime)
    fmt.Printf("Freeing allocated memory for streams on server:\t%.5fs\n", currentTime)

    // Free allocated maxfile data
    startTime = time.Now()
    client.LMemLoopbackFree()
    currentTime = calcTime(startTime)
    fmt.Printf("Freeing allocated maxfile data:\t\t\t%.5fs\n", currentTime)

    // Close!
    startTime = time.Now()
    transport.Close()
    currentTime = calcTime(startTime)
    fmt.Printf("Closing connection:\t\t\t\t%.5fs\n", currentTime)

    return outData
}

func main() {
    // Generating input data
    startTime := time.Now()
    size := int32(384)
    inA := make([]int32, size)
    inB := make([]int32, size)

    for i := int32(0); i < size; i++ {
        inA[i] = i
        inB[i] = size - i
    }
    currentTime := calcTime(startTime)
    fmt.Printf("Generating input data:\t\t\t\t%.5fs\n", currentTime)

    // DFE Output
    startTime = time.Now()
    dataOutDFE := LMemLoopbackDFE(int64(size), inA, inB);
    currentTime = calcTime(startTime)
    fmt.Printf("DFE LMemLoopback total time:\t\t\t%.5fs\n", currentTime)

    // CPU Output
    startTime = time.Now()
    dataOutCPU := LMemLoopbackCPU(int64(size), inA, inB);
    currentTime = calcTime(startTime)
    fmt.Printf("CPU LMemLoopback total time:\t\t\t%.5fs\n", currentTime)

    // Checking results
    startTime = time.Now()
    status := check(dataOutDFE, dataOutCPU, int64(size))
    currentTime = calcTime(startTime)
    fmt.Printf("Checking results:\t\t\t\t%.5fs\n", currentTime)

    if status == 0 {
        fmt.Printf("Test passed!\r\n")
    } else {
        fmt.Printf("Test failed %v times!", status)
        os.Exit(-1)
    }
}

File: examples/LMemLoopback/client/csharp/AdvancedStatic/LMemLoopbackClient.cs
// <copyright file="LMemLoopbackClient.cs" company="Maxeler">
// Copyright Maxeler. All rights reserved.
// </copyright>
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using System.Net.Sockets;

using com.maxeler.LMemLoopback;

using Thrift;
using Thrift.Protocol;
using Thrift.Transport;

/// <summary> 
/// LMemLoopback BasicStatic example
/// </summary>
internal class LMemLoopbackClient
{
    /// <summary> Checks if lMemLoopbackDfe and lMemLoopbackCpu return the same value 
    /// </summary>
    /// <param name = "dataOutDFE" > Data output from DFE </param>
    /// <param name = "dataOutCPU" > Data output from CPU </param>
    /// <param name = "size" > Size of array </param>
    /// <returns> Number of elements that doesn't match </returns>
    public static int Check(List<int> dataOutDFE, List<int> dataOutCPU, int size)
    {
        int status = 0;

        for (int i = 0; i < size; i++) 
        {
            if (dataOutDFE[i] != dataOutCPU[i]) 
            {
                Console.WriteLine("Output data @ {0} = {1} (expected {2})", i, dataOutDFE[i], dataOutCPU[i]);
                status++;
            }
        }

        return status;  
    }

    /// <summary> LMemLoopback on CPU </summary>
    /// <param name = "size"> Size of arrays </param>
    /// <param name = "inA"> First array </param>
    /// <param name = "inB"> Second array </param>
    /// <returns> Data output </returns>
    public static List<int> LMemLoopbackCPU(int size, List<int> inA, List<int> inB)
    {
        List<int> dataOut = new List<int>();

        for (int i = 0; i < size; i++)
        {
            dataOut.Add(inA[i] + inB[i]);
        }

        return dataOut;
    }

    /// <summary> LMemLoopback on DFE </summary>
    /// <param name = "size"> Size of arrays </param>
    /// <param name = "inA"> First array </param>
    /// <param name = "inB"> Second array </param>
    /// <returns> Data output </returns>
     public static List<int> LMemLoopbackDFE(int size, List<int> inA, List<int> inB)
{
        Stopwatch sw = new Stopwatch();
        List<int> outData = new List<int>();
        int sizeBytes = size * 4;
        try
        {
            // Make socket
            sw.Start();
            var transport = new TSocket("localhost", 9090);

            // Wrap in a protocol
            var protocol = new TBinaryProtocol(transport);

            // Create a client to use the protocol encoder
            var client = new LMemLoopbackService.Client(protocol);
            sw.Stop();
            Console.WriteLine("Creating a client:\t\t\t\t{0}s", sw.Elapsed.TotalMilliseconds / 1000);
            
            // Connect!
            sw.Reset(); 
            sw.Start();
            transport.Open();
            sw.Stop();
            Console.WriteLine("Opening connection:\t\t\t\t{0}s", sw.Elapsed.TotalMilliseconds / 1000);

            // Initialize maxfile
            sw.Reset(); 
            sw.Start();
            var maxfile = client.LMemLoopback_init();
            sw.Stop();
            Console.WriteLine("Initializing maxfile:\t\t\t\t{0}s", sw.Elapsed.TotalMilliseconds / 1000);

            // Load DFE
            sw.Reset(); 
            sw.Start();
            var engine = client.max_load(maxfile, "*");
            sw.Stop();
            Console.WriteLine("Loading DFE:\t\t\t\t\t{0}s", sw.Elapsed.TotalMilliseconds / 1000);

            // Allocate and send input streams to server
            sw.Reset(); 
            sw.Start();
            var address_inA = client.malloc_int32_t(size);
            client.send_data_int32_t(address_inA, inA);

            var address_inB = client.malloc_int32_t(size);
            client.send_data_int32_t(address_inB, inB);
            sw.Stop();
            Console.WriteLine("Sending input data:\t\t\t\t{0}s", sw.Elapsed.TotalMilliseconds / 1000);

            // Allocate memory for output stream on server
            sw.Reset(); 
            sw.Start();
            var address_outData = client.malloc_int32_t(size);
            sw.Stop();
            Console.WriteLine("Allocating memory for output stream on server:\t{0}s", sw.Elapsed.TotalMilliseconds / 1000);

            // Writing to LMem
            sw.Reset(); 
            sw.Start();
            var actions_lmem_write = new LMemLoopback_writeLMem_actions_t_struct();

            actions_lmem_write.Param_address = 0;
            actions_lmem_write.Param_nbytes = sizeBytes;
            actions_lmem_write.Instream_cpu_to_lmem = address_inA;

            var address_actions_lmem_write = client.send_LMemLoopback_writeLMem_actions_t(actions_lmem_write);
            client.LMemLoopback_writeLMem_run(engine, address_actions_lmem_write);

            actions_lmem_write.Param_address = sizeBytes;
            actions_lmem_write.Param_nbytes = sizeBytes;
            actions_lmem_write.Instream_cpu_to_lmem = address_inB;      

            address_actions_lmem_write = client.send_LMemLoopback_writeLMem_actions_t(actions_lmem_write);
            client.LMemLoopback_writeLMem_run(engine, address_actions_lmem_write);
            client.free(address_actions_lmem_write);
            sw.Stop();
            Console.WriteLine("Writing to LMem:\t\t\t\t{0}s", sw.Elapsed.TotalMilliseconds / 1000);

            // Action default 
            sw.Reset(); 
            sw.Start();
            var actions = new LMemLoopback_actions_t_struct();

            actions.Param_N = size;

            var address_actions = client.send_LMemLoopback_actions_t(actions);
            client.LMemLoopback_run(engine, address_actions);
            client.free(address_actions);
            sw.Stop();
            Console.WriteLine("LMemLoopback time:\t\t\t\t{0}s", sw.Elapsed.TotalMilliseconds / 1000);

            // Reading from LMem
            sw.Reset(); 
            sw.Start();
            var actions_lmem_read = new LMemLoopback_readLMem_actions_t_struct();

            actions_lmem_read.Param_address = 2 * sizeBytes;
            actions_lmem_read.Param_nbytes = sizeBytes;
            actions_lmem_read.Outstream_lmem_to_cpu = address_outData;  

            var address_actions_lmem_read = client.send_LMemLoopback_readLMem_actions_t(actions_lmem_read);
            client.LMemLoopback_readLMem_run(engine, address_actions_lmem_read);
            client.free(address_actions_lmem_read);
            sw.Stop();
            Console.WriteLine("Reading from LMem:\t\t\t\t{0}s", sw.Elapsed.TotalMilliseconds / 1000);

            // Unload DFE
            sw.Reset(); 
            sw.Start();
            client.max_unload(engine);
            sw.Stop();
            Console.WriteLine("Unloading DFE:\t\t\t\t\t{0}s", sw.Elapsed.TotalMilliseconds / 1000);

            // Get output stream from server
            sw.Reset(); 
            sw.Start();
            outData = client.receive_data_int32_t(address_outData, size);
            sw.Stop();
            Console.WriteLine("Getting output stream:\t(size = {0} bit)\t{1}s", size * 32, sw.Elapsed.TotalMilliseconds / 1000);

            // Free allocated memory for streams on server
            sw.Reset(); 
            sw.Start();
            client.free(address_inA);
            client.free(address_inB);
            client.free(address_outData);
            sw.Stop();
            Console.WriteLine("Freeing allocated memory for streams on server:\t{0}s", sw.Elapsed.TotalMilliseconds / 1000);

            // Free allocated maxfile data
            sw.Reset(); 
            sw.Start();
            client.LMemLoopback_free();
            sw.Stop();
            Console.WriteLine("Freeing allocated maxfile data:\t\t\t{0}s", sw.Elapsed.TotalMilliseconds / 1000);

            // Close!
            sw.Reset(); 
            sw.Start();
            transport.Close();
            sw.Stop();
            Console.WriteLine("Closing connection:\t\t\t\t{0}s", sw.Elapsed.TotalMilliseconds / 1000);
        }
        catch (SocketException e)
        {
            Console.WriteLine("Could not connect to the server: {0}.", e.Message);
            Environment.Exit(-1);
        }
        catch (Exception e)
        {
            Console.WriteLine("An error occured: {0}", e.Message);
            Environment.Exit(-1);
        }

        return outData;
    }

    /// <summary> Calculates LMemLoopbackCPU and LMemLoopbackDFE
    /// and checks if they return the same value. 
    /// </summary>
    /// <param name = "args"> Command line arguments </param>
    public static void Main(string[] args)
        {
        Stopwatch sw = new Stopwatch();
        int status;

        // Generate input data
        sw.Start();
        const int SIZE = 384;
        List<int> inA = new List<int>();
        List<int> inB = new List<int>();

        for (int i = 0; i < SIZE; i++)
        {
            inA.Add(i);
            inB.Add(SIZE - i);
        }

        sw.Stop();
        Console.WriteLine("Generating input data:\t\t\t\t{0}s", sw.Elapsed.TotalMilliseconds / 1000);

        // DFE Output
        sw.Reset();
        sw.Start();
        List<int> dataOutDFE = LMemLoopbackDFE(SIZE, inA, inB);
        sw.Stop();
        Console.WriteLine("LMemLoopback DFE total time:\t\t\t{0}s", sw.Elapsed.TotalMilliseconds / 1000);

        // CPU Output
        sw.Reset();
        sw.Start();
        List<int> dataOutCPU = LMemLoopbackCPU(SIZE, inA, inB);
        sw.Stop();
        Console.WriteLine("LMemLoopback CPU total time:\t\t\t{0}s", sw.Elapsed.TotalMilliseconds / 1000);

        // Checking results
        sw.Reset();
        sw.Start();
        status = Check(dataOutDFE, dataOutCPU, SIZE);
        sw.Stop();
        Console.WriteLine("Checking results:\t\t\t\t{0}s", sw.Elapsed.TotalMilliseconds / 1000);

        if (status > 0)
        {
            Console.WriteLine("Test failed {0} times! ", status);
            Environment.Exit(-1);
        } 
        else 
        {
            Console.WriteLine("Test passed!");
        }
    }
}
